### PR TITLE
fix(qdrant): add URL prefix handling for QdrantClient initialization

### PR DESCRIPTION
**Pull request by CW-B-W**
### Related GitHub Issue

Closes: #5032

Potentially (along with #4992) Resolves: #4092

### Description

This PR fixes a bug in the `QdrantVectorStore` constructor where the `QdrantClient` was not correctly configured to handle URLs with path prefixes. Previously, the `prefix` parameter was not being extracted from the URL's pathname, leading to connection issues for certain Qdrant deployments.

The fix involves:
- Modifying the `QdrantClient` instantiation in `src/services/code-index/vector-store/qdrant-client.ts` to extract the `pathname` from the parsed URL and pass it as the `prefix` option to the `QdrantClient` constructor, but only if the pathname is not the root (`/`).

### Test Procedure

**Manual Verification (Realistic Scenarios):**
I have tested with the following type of URLs:
* `https://example.com/qdrant`
* `https://example.com`
* `https://example.com:6333`
* `http://example.com`
* `http://example.com:6333`

**Unit Tests (Mocked Environment):**
New unit test cases have been added to `src/services/code-index/vector-store/__tests__/qdrant-client.spec.ts` under a new "URL Prefix Handling" describe block. These tests verify:
- That the `prefix` is correctly passed to `QdrantClient` when the URL has a non-root pathname (e.g., `http://localhost:6333/some/path`).
- That the `prefix` is `undefined` when the URL has a root pathname (e.g., `http://localhost:6333/`).
- That HTTPS URLs with paths also correctly set the prefix.

To run the unit tests:
1. Navigate to the `src/` directory: `cd src/`
2. Execute the test command: `npm test`
All unit tests, including the newly added ones, should pass.

### Pre-Submission Checklist

- [x] **Issue Linked**: This PR is linked to an approved GitHub Issue (see "Related GitHub Issue" above).
- [x] **Scope**: My changes are focused on the linked issue (one major feature/fix per PR).
- [x] **Self-Review**: I have performed a thorough self-review of my code.
- [x] **Testing**: New and/or updated tests have been added to cover my changes (if applicable).
- [x] **Documentation Impact**: I have considered if my changes require documentation updates (see "Documentation Updates" section below).
- [x] **Contribution Guidelines**: I have read and agree to the [Contributor Guidelines](/CONTRIBUTING.md).

### Screenshots / Videos

N/A (Backend logic change)

### Documentation Updates

- [x] No documentation updates are required.

### Additional Notes

None.

### Get in Touch

----

> [!IMPORTANT]
> Fixes URL prefix handling in `QdrantVectorStore` by correctly setting `prefix` in `QdrantClient` initialization.
> 
>   - **Behavior**:
>     - Fixes URL prefix handling in `QdrantVectorStore` constructor in `qdrant-client.ts` by extracting `pathname` from URL and passing it as `prefix` to `QdrantClient` if not root.
>   - **Tests**:
>     - Adds tests in `qdrant-client.spec.ts` to verify `prefix` is set correctly for non-root path URLs and is `undefined` for root path URLs.
>     - Tests HTTPS URLs with paths to ensure correct prefix setting.
> 
> <sup>This description was created by </sup>[Ellipsis](https://www.ellipsis.dev?ref=RooCodeInc%2FRoo-Code&utm_source=github&utm_medium=referral)<sup> for 9b09f08d95fec4e5b1a7ec1551b3391fe6d57b34. You can [customize](https://app.ellipsis.dev/RooCodeInc/settings/summaries) this summary. It will automatically update as commits are pushed.</sup>

### CODE REVIEW

File: src/services/code-index/vector-store/qdrant-client.ts [hunk 1]
```diff
@@ -57,13 +57,15 @@ export class QdrantVectorStore implements IVectorStore {
 				host: urlObj.hostname,
 				https: useHttps,
 				port: port,
+				prefix: urlObj.pathname === "/" ? undefined : urlObj.pathname,
```
Comment (daniel-lxs) [inline src/services/code-index/vector-store/qdrant-client.ts]: I noticed that URLs with trailing slashes aren't normalized. Would `http://localhost:6333/api/` and `http://localhost:6333/api` result in different prefix values (`/api/` vs `/api`)? If so, should we consider normalizing the pathname to ensure consistent behavior?

For example:
```typescript
prefix: urlObj.pathname === "/" ? undefined : urlObj.pathname.replace(/\/$/, '')
```

File: src/services/code-index/vector-store/qdrant-client.ts [hunk 2]
```diff
@@ -57,13 +57,15 @@ export class QdrantVectorStore implements IVectorStore {
...
 				apiKey,
 				headers: {
 					"User-Agent": "Roo-Code",
 				},
 			})
 		} catch (urlError) {
 			// If URL parsing fails, fall back to URL-based config
+			// Note: This fallback won't correctly handle prefixes, but it's a last resort for malformed URLs.
```
Comment (daniel-lxs) [inline src/services/code-index/vector-store/qdrant-client.ts:68]: The comment mentions that this fallback won't correctly handle prefixes. Could this lead to silent failures for users with malformed URLs that require path prefixes? Would it be worth logging a warning here or throwing an error to make the issue more visible?

File: src/services/code-index/vector-store/__tests__/qdrant-client.spec.ts [hunk 3]
```diff
@@ -337,6 +343,65 @@ describe("QdrantVectorStore", () => {
...
+				port: 443,
+				prefix: "/api",
+				apiKey: undefined,
+				headers: {
+					"User-Agent": "Roo-Code",
+				},
+			})
+			expect((vectorStoreWithHttpsPrefix as any).qdrantUrl).toBe("https://qdrant.ashbyfam.com/api")
+		})
+	})
```
Comment (daniel-lxs) [inline src/services/code-index/vector-store/__tests__/qdrant-client.spec.ts:497]: Would it be helpful to add test cases for these edge cases?
- URLs with trailing slashes (e.g., `/api/` vs `/api`)
- Multiple path segments (e.g., `/api/v1/qdrant`)
- URLs with query parameters or fragments (e.g., `/api?key=value` or `/api#section`)

These tests could help ensure the prefix handling is robust across different URL formats.

**Review by daniel-lxs — COMMENTED** [review]
Thanks for this fix! The implementation looks good. I've left a few suggestions for edge cases and error handling that might be worth considering.

File: src/services/code-index/vector-store/qdrant-client.ts (hunk 1 above)
Comment (CW-B-W) [reply src/services/code-index/vector-store/qdrant-client.ts]: Appreciate for you suggestion, I apprently ignored that.
I have revised it and committed.

File: src/services/code-index/vector-store/__tests__/qdrant-client.spec.ts (hunk 3 above)
Comment (CW-B-W) [reply src/services/code-index/vector-store/__tests__/qdrant-client.spec.ts:497]: Thanks for your suggestion!

I have added more test cases.

Also for those cases, I tried them under realistic scenarios
* https://example.com/qdrant/
* https://example.com/qdrant///
* https://example.com/service/api/qdrant/
* https://example.com/service/api/qdrant///
* https://example.com/service/api/qdrant/?key=value#section
* https://example.com/service/api/qdrant///?key=value#section

They all worked

**Review by mrubens — APPROVED** [review]

**Comment by hannesrudolph** [comment]
@CW-B-W If you have discord would you be able to message me (hrudolph)?
